rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,356 @@
+
+- CSP(Cloud Storage Provider) - Cloud storage services such as Dropbox or Google Drive.
+- Shared Link - A link from a CSP that give the bearer of that link access to private data.
+- DP(Data Portal) - The cellxgene Data Portal.
+- Small File - a file < 5 GB in size.
+- Large File - a file > 5 GB in size.
+
+## Problem Statement | Background
+
+One of the primary goals of the DP is to allow users to upload their own data sets. These data sets can vary in
+size from a few MB to several GB. User may want to upload from their local machine or from their CSP. For
+users that want to upload from their local machine, they must be able to resume an upload in the event their connection
+is interrupted. For users that want to upload data sets from their CSP, they are able to provide a shared link from
+their CSP that is used to upload directly to S3.
+
+## Product Requirements
+
+1. A user can upload a file to DP using a shared link from their CSP.
+1. A user can upload a file from their local machine using the DP browser.
+1. A user can resume an upload without completely restarting.
+1. The user must be authenticated before uploading.
+1. The publisher must also agree to portal policies as outlined in
+   [Publishers must complete attestation before submitting private datasets to the portal](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/22).
+
+### Nonfunctional requirements
+
+Nonfunctional requirements are not directly related to the user story but may instead reflect some technical aspect of the implementation, such as latency. [Here's](https://en.wikipedia.org/wiki/Non-functional_requirement) a more comprehensive list of nonfunctional requirements.
+
+1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.
+1. A user can upload a single file up to 30GB in size.
+
+## Detailed Design | Architecture | Implementation
+
+![Upload Design](https://app.lucidchart.com/publicSegments/view/a74f747f-d071-41ea-949f-52f38a98747a/image.png)
+Figure 1: The Upload Design
+
+### Components
+
+This section describes the purpose of components that make up the upload design.
+
+#### Data Portal Browser App
+
+The DP Browser App is the frontend of the system. It is the primary method users have to interact with the Data Portal.
+It will be used authenticate the user, and interact with the DP Backend to upload files. It will also use AWS JS SDK to
+upload files from local storage to AWS. These feature will extend the existing DP Frontend.
+
+#### Data Portal Backend
+
+The DP Backend is responsible for handling users requests. It will generated presigned URL and access tokens
+for the DP Browser App to facilitate the uploading of files. It will also enqueue shared links for downloading.
+This additional functionality can be built into the existing DP Backend Lambda. The backend will also be responsible for
+performing quick validation of the filing being uploaded
```

Comment:
```suggestion
performing quick validation of the file being uploaded
```

Reply:
I think in practice there isn't a need to say "quick" validation. The full blown validation will be performed on every file that is uploaded.

Reply:
this a validation that occurs on the frontend before the file is uploaded. I think this is actually outside the scope of the Design.